docs/filesystem_merkletree.md

```diff
@@ -0,0 +1,188 @@
+and then the function will set `leaves_inine`. (same logic, but different function implementation for `DirectoryNode.contents_inline` vs
+`DirectoryNode.contents_external`)
+
+It aims for 16-64kb for objects holding directory information, and 64-256kb for objects holding files.
+
+This inlining/externalizing process happens at every level of the filetree.
+
+At the end of the day the original caller just gets a Directory message back, and can get to work.
+
+Interesting observations:
+- Objects that hold the files will be relatively stable. If the file doesn't change much, the objects that file is spread over don't change much. The closer you get to the root of a tree,
+say the linux kernel tree, the more the objects will churn. Any change in a file will cause the object that holds the Directory
+that references it to also change, and the object that holds the Directory that points to that also change.
+
+- You can only reference an object as a whole. There's no way to index into it. Therefore the only things, in the context of this filesystem work, you can store in
+  objects are:
+     - single protobuf (like Leaves)  
+     - single array of protobufs (like array of FileContents)
+     - raw bytes for a file
+
+Let's take a closer look at how files are dealt with. Above I brushed over it when I wrote that "you make a Leaves message for the files and symlinks".
+When the algorithm above encounters a file that's not very big, then it's easy: it will make a single object for the file, 
+and the Leaves message will be populated with a FileContents message that points to that object.
+
+When the file is big, things get more interesting. It will be cut up into small pieces. Each piece goes into its own object,
+and there will be another "indirection" object that holds an array of FileContents messages that point to all the segment objects.
+Finally the Leaves message will have a FileContents message that points to the indirection object.
+
+The art is in deciding how to chop it up. If you chop it up naively, say in 256kb segments, a tiny change in the beginning
+of the file, will cause all objects to change. Turns out there's smarter ways to cut it up. If you somehow would be able to
+find the same cut points as last time, even if there have been some changes here and there, many segments could remain identical.
+We use a trick where we look at a sliding window of 64 bytes at a time, generate a deterministic random number from them (by hashing them), and then
+saying "if the number is high, this is a good cut point". It's completely arbitrary, but that's okay, as long as next time you
+use the same heuristic to find cut points, you are likely to find the same cut points as last time, and you'll get many identical
+segment objects, even in the face of changes having occurred in the file.
+
+When the file is really big, a single indirection object is not enough. Objects have a max of 256 outgoing references.
```

Comment:
That's not entirely correct. Right now they can have a max of 2 MiB / 40 B = 52428 outgoing references. But in practice it's hard to reach that, because you also want to store a data payload that tie those objects together.

Reply:
Relatedly, there is a limit of 256 on the height of the Merkle tree.